streams/src/main/java/org/apache/kafka/streams/state/internals/TimestampedKeyValueStoreBuilder.java

```diff
@@ -48,8 +53,12 @@ public TimestampedKeyValueStoreBuilder(final KeyValueBytesStoreSupplier storeSup
     @Override
     public TimestampedKeyValueStore<K, V> build() {
         KeyValueStore<Bytes, byte[]> store = storeSupplier.get();
-        if (!(store instanceof TimestampedBytesStore) && store.persistent()) {
-            store = new KeyValueToTimestampedKeyValueByteStoreAdapter(store);
+        if (!(store instanceof TimestampedBytesStore)) {
+            if (store.persistent()) {
+                store = new KeyValueToTimestampedKeyValueByteStoreAdapter(store);
+            } else {
+                store = new InMemoryTimestampedKeyValueStoreMarker(store);
+            }
```

Comment:
This fixes a bug that was exposed after rebasing. We need to treat in-memory stores always as `TimestampedByteStores` to guarantee correct restore. (Similar below for window stores.)

Reply:
Added a new test to #6179 that also exposed this bug (cf. `StoreUpgradeIntegrationTest`)

Reply:
I agree with the fix, but it raises another question for me.

I admit I could be either missing something or forgetting some details already spelled out.  Why don't we convert the values for in-memory stores to also have timestamps in the value?

Reply:
The in-memory store will have no content on startup, and the upper layer will put/get `byte[]` -- the store does not care what is in it. The upper layer will already serialize value+timestamp and put those, and expects value+timestamp `byte[]` back -- hence, the in-memory store does not need to do any translation.

For persistent store, the store does have content on local disk, namely plain values. Because this store cannot be upgraded, we put the adapter in between, to remove the timestamp part on put() and add dummy timestamp on get(), to meet the contract to the upper layer that expects `byte[]` to be value+timestamp.

If we restore a persistent store, the just copy plain key/value from the changelog topic and put it into the store. Everything still works as expected.

If we restore an in-memory store, we need to add the timestamp to the value, and this is triggered by the `TimestampeByteStore` interface. Otherwise, the in-memory store would contain plain values, and violate the contract to the upper layer, because the assumption is, that everything store in it, has value+timestamp format.

Ie, for persistent store, we play the backward compatibility game using surrogate timestamps. However, for in-memory store, we don't need to do this, because there is no local on-disk content and we can switch to timestamped-store behavior without any drawbacks and get full feature set.

Does this make sense?

Reply:
yep thanks for clarifying